src/main/java/org/wso2/carbon/inbound/kafka/deserializer/ErrorHandlingDeserializer.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (c) 2023, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 LLC. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wso2.carbon.inbound.kafka.deserializer;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.header.Headers;
+import org.apache.kafka.common.header.internals.RecordHeader;
+import org.apache.kafka.common.header.internals.RecordHeaders;
+import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.synapse.SynapseException;
+import org.wso2.carbon.inbound.kafka.KafkaConstants;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectOutputStream;
+import java.util.Map;
+
+public class ErrorHandlingDeserializer<T> implements Deserializer<T> {
+
+    private static final Log LOGGER = LogFactory.getLog(ErrorHandlingDeserializer.class);
+
+    private Deserializer<T> delegate;
+
+    private boolean isForKey;
+
+    public ErrorHandlingDeserializer() {
+    }
+
+    public ErrorHandlingDeserializer(Deserializer<T> delegate) {
+        this.delegate = setupDelegate(delegate);
+    }
+
+    @Override
+    public void configure(Map<String, ?> configs, boolean isKey) {
+
+        if (this.delegate == null) {
+            setupDelegate(configs, isKey ? KafkaConstants.KEY_DELEGATE_DESERIALIZER
+                    : KafkaConstants.VALUE_DELEGATE_DESERIALIZER);
+        }
+        if (this.delegate == null) {
+            LOGGER.error("No delegate deserializer configured");
+        }
+        this.delegate.configure(configs, isKey);
+        this.isForKey = isKey;
+    }
+
+    @Override
+    public T deserialize(String topic, byte[] data) {
+
+        try {
+            return this.delegate.deserialize(topic, data);
+        } catch (Exception e) {
```

Comment:
Shouldn't we set the deserialization headers only for serialization related exceptions?

Reply:
No, we need to add the headers when there is an error while deserializing the data. The actual cause can be anything that causes issues in the deserialize method.